JDK-8302644: IGV: Apply filters per graph tab and not globally

tobiasholenstein opened the pull request:
In IGV the user can apply a set of filters to a graph. Currently, the same set of selected filters is applied to all graphs (globally). 

- With this change the use can define a set of filters for each individual graph tab using the `--Local--` profile
- Further a filter profile can be created that represents a set of filter. This filter profile can the be selected in each graph tab individually.

### Global profile
Each tab has a  `--Global--` filter profile which is selected when opening a graph. Filters applied to the `--Global--` profile are applied to all tabs that have the `--Global--` profile selected. 

### Local profile
Each tab has its own  `--Local--` filter profile. Filters applied to the `--Local--` profile are applied only to the currently selected tabs. Only one tab can be selected at a time and a tab gets selected by clicking on it. To make it more clear which tab is currently selected, the title of the selected tab is displayed in **bold** font.
<img width="1252" alt="tabA" src="https://user-images.githubusercontent.com/71546117/222127949-38a3ed8f-8f61-4485-aca6-67cbc341a4d1.png">

When clicking on a different tab with a different `--Local--` profile, the selected filters get updated accordingly.
<img width="1256" alt="tabB" src="https://user-images.githubusercontent.com/71546117/222127988-66be7a76-a638-4c40-b8f1-ad56a1ab1fd7.png">

### New profile
The user can also create a new filter profile and give it a name. E.g. `My Filters`
<img width="396" alt="newProfile" src="https://user-images.githubusercontent.com/71546117/222128202-ea1f8de6-9260-4a78-b396-b69a354dc883.png">

The `My Filters` profile is then globally available to other tabs as well 
<img width="254" alt="selectProfile" src="https://user-images.githubusercontent.com/71546117/222128229-e33ab7b1-ebbe-4abc-a759-b50965ca64e1.png">


### Filters for cloned tabs
When the user clones a tab, the `--Local--` profile gets cloned as well. Further the clone has the same filter profile selected when it gets opened
<img width="1250" alt="cloneTab" src="https://user-images.githubusercontent.com/71546117/222128014-2ca5eac0-ae8a-42da-b407-acc62c09edcf.png">

### Saving of filters and profiles
When the users closes IGV, the filters (in their exact order) are save, as well as the filter profiles. The profile that was last used is selected when opening IGV.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302644](https://bugs.openjdk.org/browse/JDK-8302644): IGV: Apply filters per graph tab and not globally


### Reviewers
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**) ⚠️ Review applies to [2d6409b9](https://git.openjdk.org/jdk/pull/12714/files/2d6409b9590ad7e45579db2976ebc770ef88c706)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [ba9aacc3](https://git.openjdk.org/jdk/pull/12714/files/ba9aacc3c31279bc783a38f90b34197513c993c2)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [ba9aacc3](https://git.openjdk.org/jdk/pull/12714/files/ba9aacc3c31279bc783a38f90b34197513c993c2)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12714/head:pull/12714` \
`$ git checkout pull/12714`

Update a local copy of the PR: \
`$ git checkout pull/12714` \
`$ git pull https://git.openjdk.org/jdk.git pull/12714/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12714`

View PR using the GUI difftool: \
`$ git pr show -t 12714`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12714.diff">https://git.openjdk.org/jdk/pull/12714.diff</a>

</details>

robcasloz commented:
Thanks for working on this, Toby! Being able to apply different filters per tab is definitely useful feature. However, as an IGV user I miss two things from the current behavior: persistence (the same filters are applied after restarting IGV) and the ability to apply the same filter configuration to all tabs in a simple manner. 

I would like to propose an alternative model that is almost a superset of what is proposed here and would preserve persistence and easy filter synchronization among tabs. By default, each tab has two filter profiles available, "local" and "global". More profiles cannot be added or removed. The local filter profile can be edited but is not persistent (i.e. it acts like the `--Custom--` profile in this changeset). The global filter profile can be edited, is persistent, and the changes are propagated for all tabs where it is selected. The `Link node selection globally` button is generalized to `Link node and filter selection globally`. It is disabled by default, and clicking on it selects the global filter profile for all opened tabs.

What do you think? This is just my input as user, it would be useful to see what others think here.

tobiasholenstein commented:
(I renamed `--Custom--` to `--Local--`)

robcasloz commented:
> I now added also a --Global-- profile that is selected by default.

Thanks for the changes, Toby. I can see the `--Global--` profile selected by default, however as soon as I open a graph it switches to `--Local--`. Is this intended?

robcasloz commented on src/utils/IdealGraphVisualizer/FilterWindow/src/main/java/com/sun/hotspot/igv/filterwindow/FilterTopComponent.java:
```diff
@@ -472,20 +447,18 @@ public void initFilters() {
             map.put(displayName, cf);
 
             String after = (String) fo.getAttribute(AFTER_ID);
+            System.out.println(displayName + " after " + after);
```
Please remove.

tobiasholenstein replied:
done

robcasloz commented on src/utils/IdealGraphVisualizer/FilterWindow/src/main/java/com/sun/hotspot/igv/filterwindow/FilterTopComponent.java:
```diff
@@ -626,72 +565,81 @@ public void requestActive() {
+        }
+        allFiltersOrdered.sortBy(order);
+
+        System.out.println("readExternal");
```
Please remove.

robcasloz commented on src/utils/IdealGraphVisualizer/Filter/src/main/java/com/sun/hotspot/igv/filter/FilterChain.java:
```diff
@@ -37,60 +38,44 @@
-        for (Filter f : filters) {
-            if (!applied.contains(f)) {
-                f.apply(d);
+    public void applyInOrder(Diagram diagram, FilterChain filterOrder) {
```
I have also addressed this in [#12955](https://github.com/openjdk/jdk/pull/12955), but I think your solution of sorting the filter list upfront, rather than every time filters are applied, is preferable. I will wait for this PR to be integrated and then exclude the corresponding changes from #12955.

tobiasholenstein replied:
sounds good

robcasloz commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/DiagramViewModel.java:
```diff
@@ -290,15 +291,50 @@ public void setHiddenNodes(Set<Integer> nodes) {
     }
 
+    // called when the filter in filterChain changed, but not filterChain itself
     private void filterChanged() {
```
After applying this PR, `DiagramViewModel::filterChanged()` is fired every time a new graph in a group is viewed. This is not a functional bug, but it causes the expensive `DiagramViewModel::rebuildDiagram()` to be called twice in that scenario (the other call comes from `DiagramViewModel::changed()`). Would it be possible to arrange the code so that `DiagramViewModel::rebuildDiagram()` is only called once when a new graph in a group is viewed?

tobiasholenstein replied:
You are right. `DiagramViewModel::rebuildDiagram()` was called too many times when it was not necessary. I updated the code - now `DiagramViewModel::rebuildDiagram()` should be only called when needed

tobiasholenstein replied:
my changed caused the checkbox to not be updated anymore when changing the profile (thanks @chhagedorn for finding that bug). Should be fixed now.

robcasloz reviewed (changes requested):
Thanks for addressing my comments again, Toby! I like the additional functionality and that the changeset preserves the current workflow by default. I just have a question about graph viewing performance and a couple of minor comments.

robcasloz reviewed (approved):
Looks good!

chhagedorn commented on src/utils/IdealGraphVisualizer/Filter/src/main/java/com/sun/hotspot/igv/filter/FilterChain.java:
```diff
@@ -129,7 +122,21 @@ public void moveFilterDown(Filter filter) {
+
     public List<Filter> getFilters() {
         return Collections.unmodifiableList(filters);
     }
```
Missing new line
```suggestion
    }

```

chhagedorn commented on src/utils/IdealGraphVisualizer/FilterWindow/src/main/java/com/sun/hotspot/igv/filterwindow/FilterTopComponent.java:
```diff
@@ -327,80 +342,46 @@ private static String getJsHelperText() {
 
-    public void removeFilter(Filter f) {
-        com.sun.hotspot.igv.filter.CustomFilter cf = (com.sun.hotspot.igv.filter.CustomFilter) f;
+    public void addFilter(CustomFilter cf) {
```
While at it, you could rename `cf` and `fo` to something more descriptive like `customFilter` and `fileObject`.

tobiasholenstein replied:
done

chhagedorn commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/EditorTopComponent.java:
```diff
@@ -372,9 +366,38 @@ public void requestActive() {
+        }
+    }
+
+    private boolean useBoldDisplayName = false;
```
Should be moved up to the other fields

chhagedorn commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/EditorTopComponent.java:
```diff
@@ -372,9 +366,38 @@ public void requestActive() {
+        useBoldDisplayName = bold;
+        setDisplayName(getDisplayName());
+    }
+
+
```
```suggestion

```

chhagedorn reviewed (approved):
That's a nice improvement! I've tried your patch out again and it works as expected. Thanks for fixing the issue discussed offline. I only have some minor code style comments. Otherwise, looks good to me!

tobiasholenstein commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/DiagramViewModel.java:
```diff
@@ -147,47 +157,69 @@ public boolean getHideDuplicates() {
         showCFG = Settings.get().getInt(Settings.DEFAULT_VIEW, Settings.DEFAULT_VIEW_DEFAULT) == Settings.DefaultView.CONTROL_FLOW_GRAPH;
         showNodeHull = true;
         showEmptyBlocks = true;
-        group = graph.getGroup();
```
good idea - done

tobiasholenstein commented on src/utils/IdealGraphVisualizer/Util/src/main/java/com/sun/hotspot/igv/util/RangeSliderModel.java:
```diff
@@ -45,7 +45,17 @@ public class RangeSliderModel implements ChangedEventProvider<RangeSliderModel>
     private List<Color> colors;
 
+    public RangeSliderModel(RangeSliderModel model) {
+        this();
```
done

TobiHartmann commented:
Works well for me but I spotted the following issue which seems to be a regression from this change:
- Open two .xml files, open a graph in each and select the local profile
- Double click on a filter and then click Cancel

```
java.lang.AssertionError
	at com.sun.hotspot.igv.view.DiagramViewModel.filterChanged(DiagramViewModel.java:356)
	at com.sun.hotspot.igv.view.DiagramViewModel.lambda$new$1(DiagramViewModel.java:73)
	at com.sun.hotspot.igv.data.ChangedEvent.fire(ChangedEvent.java:44)
	at com.sun.hotspot.igv.data.ChangedEvent.fire(ChangedEvent.java:31)
	at com.sun.hotspot.igv.data.Event.fire(Event.java:56)
	at com.sun.hotspot.igv.filter.FilterChain$1.changed(FilterChain.java:48)
	at com.sun.hotspot.igv.filter.FilterChain$1.changed(FilterChain.java:45)
	at com.sun.hotspot.igv.data.ChangedEvent.fire(ChangedEvent.java:44)
	at com.sun.hotspot.igv.data.ChangedEvent.fire(ChangedEvent.java:31)
	at com.sun.hotspot.igv.data.Event.fire(Event.java:56)
	at com.sun.hotspot.igv.filter.CustomFilter.openInEditor(CustomFilter.java:86)
	at com.sun.hotspot.igv.filter.CustomFilter$1.open(CustomFilter.java:77)
	at org.openide.actions.OpenAction.performAction(OpenAction.java:59)
	at org.openide.util.actions.NodeAction$DelegateAction$1.run(NodeAction.java:561)
	at org.openide.util.actions.ActionInvoker$1.run(ActionInvoker.java:70)
	at org.openide.util.actions.ActionInvoker.doPerformAction(ActionInvoker.java:91)
	at org.openide.util.actions.ActionInvoker.invokeAction(ActionInvoker.java:74)
	at org.openide.util.actions.NodeAction$DelegateAction.actionPerformed(NodeAction.java:558)
	at org.openide.explorer.view.ListView.performObjectAt(ListView.java:681)
	at org.openide.explorer.view.ListView$PopupSupport.mouseClicked(ListView.java:1306)
	at java.desktop/java.awt.AWTEventMulticaster.mouseClicked(AWTEventMulticaster.java:278)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6638)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
	at java.desktop/java.awt.Component.processEvent(Component.java:6400)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5011)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4843)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4556)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4843)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)
[catch] at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

tobiasholenstein commented:
> Works well for me but I spotted the following issue which seems to be a regression from this change:
> 
> * Open two .xml files, open a graph in each and select the local profile
> * Double click on a filter and then click Cancel
> 
> ```
> java.lang.AssertionError
> 	at com.sun.hotspot.igv.view.DiagramViewModel.filterChanged(DiagramViewModel.java:356)
> 	at com.sun.hotspot.igv.view.DiagramViewModel.lambda$new$1(DiagramViewModel.java:73)
> 	at com.sun.hotspot.igv.data.ChangedEvent.fire(ChangedEvent.java:44)
> 	at com.sun.hotspot.igv.data.ChangedEvent.fire(ChangedEvent.java:31)
> 	at com.sun.hotspot.igv.data.Event.fire(Event.java:56)
> 	at com.sun.hotspot.igv.filter.FilterChain$1.changed(FilterChain.java:48)
> 	at com.sun.hotspot.igv.filter.FilterChain$1.changed(FilterChain.java:45)
> 	at com.sun.hotspot.igv.data.ChangedEvent.fire(ChangedEvent.java:44)
> 	at com.sun.hotspot.igv.data.ChangedEvent.fire(ChangedEvent.java:31)
> 	at com.sun.hotspot.igv.data.Event.fire(Event.java:56)
> 	at com.sun.hotspot.igv.filter.CustomFilter.openInEditor(CustomFilter.java:86)
> 	at com.sun.hotspot.igv.filter.CustomFilter$1.open(CustomFilter.java:77)
> 	at org.openide.actions.OpenAction.performAction(OpenAction.java:59)
> 	at org.openide.util.actions.NodeAction$DelegateAction$1.run(NodeAction.java:561)
> 	at org.openide.util.actions.ActionInvoker$1.run(ActionInvoker.java:70)
> 	at org.openide.util.actions.ActionInvoker.doPerformAction(ActionInvoker.java:91)
> 	at org.openide.util.actions.ActionInvoker.invokeAction(ActionInvoker.java:74)
> 	at org.openide.util.actions.NodeAction$DelegateAction.actionPerformed(NodeAction.java:558)
> 	at org.openide.explorer.view.ListView.performObjectAt(ListView.java:681)
> 	at org.openide.explorer.view.ListView$PopupSupport.mouseClicked(ListView.java:1306)
> 	at java.desktop/java.awt.AWTEventMulticaster.mouseClicked(AWTEventMulticaster.java:278)
> 	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6638)
> 	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
> 	at java.desktop/java.awt.Component.processEvent(Component.java:6400)
> 	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
> 	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5011)
> 	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
> 	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4843)
> 	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
> 	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4556)
> 	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
> 	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
> 	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
> 	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4843)
> 	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
> 	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
> 	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
> 	at java.base/java.security.AccessController.doPrivileged(Native Method)
> 	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
> 	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
> 	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
> 	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
> 	at java.base/java.security.AccessController.doPrivileged(Native Method)
> 	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
> 	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
> 	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)
> [catch] at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
> 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
> 	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
> ```

Thanks for catching that @TobiHartmann . It should be fixed now

TobiHartmann reviewed (approved):
Works well now. Thanks for fixing!

chhagedorn commented on src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/DiagramViewModel.java:
```diff
@@ -343,21 +348,14 @@ void activateModel() {
-    }
-
     void close() {
+        System.out.println(getGraph().getDisplayName() + " removeListener to " + filterChain.getName());
```
Leftover from debugging?

tobiasholenstein replied:
right. I removed it

chhagedorn reviewed (approved):
Thanks for making the changes! The update looks good.

tobiasholenstein commented:
Thanks @robcasloz , @chhagedorn and @TobiHartmann for the reviews! 

/integrate
